ndb/src/google/cloud/ndb/_eventloop.py

```diff
@@ -141,18 +145,18 @@ def queue_call(self, delay, callback, *args, **kwargs):
         event = _Event(when, callback, args, kwargs)
         self.insort_event_right(event)
 
-    def queue_rpc(self, rpc, callback=None, *args, **kwds):
-        """Schedule an RPC with an optional callback.
-
-        The caller must have previously sent the call to the service.
-        The optional callback is called with the remaining arguments.
+    def queue_rpc(self, future, rpc):
+        """Add an RPC call.
+        """
+        rpc_id = uuid.uuid1()
+        self.rpcs[rpc_id] = future
 
-        .. note::
+        def callback(rpc):
```

Comment:
It may be overkill but you may be happier with a `__call__`-able object that wraps `self.rpc_results` and `rpc_id`:

```python
class DoneCallback:
    def __init__(self, queue, rpc_id):
        self.queue = queue
        self.rpc_id = rpc_id

    def __call__(self, rpc):
        self.queue.put((self.rpc_id, rpc))
```

I always have a hard time reading functions that aren't defined at the top level or as class methods

Reply:
Different strokes. I find the closure easier to read, personally.